x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java

```diff
@@ -90,19 +93,35 @@ public enum DataType {
      */
     private final boolean docValues;
 
-    DataType(String esName, int size, boolean isInteger, boolean isRational, boolean hasDocValues) {
-        this(null, esName, size, isInteger, isRational, hasDocValues);
-    }
+    /**
+     * {@code true} if this is a TSDB counter, {@code false} otherwise.
+     */
+    private final boolean isCounter;
+
+    /**
+     * If this is a "small" numeric type this contains the type ESQL will
+     * widen it into, otherwise this is {@code null}.
+     */
+    private final DataType widenSmallNumeric;
```

Comment:
This doesn't seem a property of the type, IMO, but a type's user preference.
If we add, say, float through support, we'd no longer widen the values to double.

Reply:
I figured we'd remove that widening in that case. It feels a little simpler to stick it here, but I'm not tightly tied to it.

Reply:
I merged this on Friday so Mark could build on this. I covered some of the followups in #110105. But not this one. I kind of like keeping this where it is....

Reply:
Was a minor note, it can stay as is 👍

Reply:
:heart: